# Welcome to Python!
Python is a full-featured programming language, but is often billed as being easier to learn than Java-like languages. You can use it with a traditional IDE that has an editor, compiler, and window to view results. However, Python is also frequently used in an environment like this one, called a **Notebook**.

Each box you see here is called a **cell**. A cell can contain just text (you are reading a text cell now) or Python code. Python can be used interactively; that is, you don't need to write a whole program, compile it, and execute it. Instead, you can write little snippets of Python code and run each one interactively. You can see the results right away rather than switching back and forth between and editor and output window.

Let's get started with some simple Python arithmetic.


# Numbers
Python supports the usual integers and floating-point numbers. However, integers can be of arbitrary size. Python happily works with integers much, much larger than other languages like Java or C natively support. It also comes with built-in support for converted to and from a variety of number bases that are useful to programmers, helping to make Python a very attrative alternative to the more traditional languages.

Python supports the usual arithmetic operators, plus some new ones that you will find useful:

| Operator | Meaning |
|---|---|
| \*, +, -, / | Multiply, add, subtract, divide |
| // | Integer division |
| % | Modulo (remainder) |
| \*\* | Exponent |

### Try it!

Try the calculations below. Click in each cell and press Shift-Enter to run them. Try changing the values and re-running them to see the results.

In [33]:
2+3+4*5

25

In [32]:
123 ** 456

99250068772098856700831462057469632637295940819886900519816298881382867104749399077921128661426144638055424236936271872492800352741649902118143819672601569998100120790496759517636465445895625741609866209900500198407153244604778968016963028050310261417615914468729918240685487878617645976939063464357986165711730976399478507649228686341466967167910126653342134942744851463899927487092486610977146112763567101672645953132196481439339873017088140414661271198500333255713096142335151414630651683065518784081203678487703002802082091236603519026256880624499681781387227574035484831271515683123742149095569260463609655977700938844580611931246495166208695540313698140011638027322566252689780838136351828795314272162111222231170901715612355701347552371530013693855379834865667060014643302459100429783653966913783002290784283455628283355470529932956051484477129333881159930212758687602795088579230431661696010232187390436601614145603241902386663442520160735566561

In [31]:
92 ** 5 % 19

4

You can also insert new cells. Just click in a cell and then press the \[+\] button in the toolbar. Use the empty cell below to perform a calculation using integer division. Then add a new cell below it to do a calculation using a variety of operators.

In [24]:
2 / 3

0.6666666666666666

In [25]:
2 * 5

10

# Entering Numbers in Other Bases
Python allows you to enter interger numbers using binary (base 2), octal (base 8), decimal (base 10), and hexadecimal (base 16). Decimal numbers are entered the usual way. For the other bases, special prefixes are used to indicate the base.

| Base | Prefix | Example |
|------|--------|---------|
| Binary | 0b | 0b1001101 |
| Octal | 0o | 0o7234 |
| Hexadecimal | 0x | 0x45a8e |

Python displays in decimal by default. When you try the following examples, you'll enter the numbers using the indicated base but the value will be shown in base 10.

### Try it!
Evaluate the numbers in the cells below and see that they are display in base 10.

Remember: you can also insert new cells. Just click in a cell and then press the [+] button in the toolbar.

In [26]:
0b1100110

102

In [27]:
0x7f3ab

521131

In [28]:
0b11111110

254

### Artithmetic
Try the following examples of doing arithmetic in other number bases.

In [29]:
0x4567 + 0b01110010011

18682

In [30]:
123456 * 0x45

8518464

# Converting To Other Bases
Python can convert integers into other bases using built-in conversion functions. The results will be displayed as strings, with quotes around them, because Python only displays numbers in base 10. Any other base must be displayed in a form that preserves the special formatting of the other bases.

In [ ]:
hex(12345)

In [ ]:
bin(651)

In [ ]:
bin(0x6fe1)

In [ ]:
hex(0b10101111)

# Arbitrary Bases
Our last examples of base conversion will demonstrate that Python can convert from any base between 2 and 36, including oddball bases like "base 3" and "base 21". The `int` function allows you to pass in a string containing the number along with what base it is.

The `int` function can also be used simply to convert a string, such as from user input, into a number.

In [37]:
int('012112', 3)

149

In [36]:
int('123ABC', 26)

12855114

In [35]:
int('6fe1', 16)

28641

In [34]:
int('123')

123

# Strings
String literals can be surrounded by either single-quotes or double-quotes. There is no difference between them. Just use whichever is more convenient. For example, `"aardvark"` and `'aardvark'` are the same strings.

Python does not distinguish between strings and single characters. A character is just a string of length 1.

Strings can be joined (concatenated) using the `+` operator, similar to Java.

### Try it!

Change the string concatenation below so there is a space between **SpongeBob** and **SquarePants**.

In [23]:
'SpongeBob' + ' ' + 'SquarePants'

'SpongeBob SquarePants'

# ASCII and Unicode values
Python can easily convert characters to and from their corresponding ASCII and Unicode values. The **ord** function is used to convert a character to Unicode.

Try the examples below by clicking in each cell and pressing Shift-Enter to execute them.

In [40]:
# The Unicode value of "A" is 65. Run this cell to verify.
ord("A")

65

In [39]:
ord("%")

37

In [38]:
ord("🙂")

128578

What is the Unicode value of the 🐍 (snake) character? (Get it? It's a Python!)

You can copy-and-paste the character into the cell below, using the `ord` function to show the Unicode value.

In [21]:
ord("🐍")

128013

# Byte Strings
Internally, strings are just a sequence of characters. (Actually, they are an array of characters. We'll learn about arrays in the coming weeks.) As we've learned above, characters have a huge range of values from small (such as 65 for the character 'A') to large (128578 for the 🙂 emoji character). Therefore, strings are really sequences of these Unicode values whose sizes can vary from a single byte to multiple bytes.

It is often useful to work with sequences of bytes, whose values can only fit into 8 bits and range from 0-255. The original ASCII character set has only 128 characters and  fits into 7 bits, with an extra bit unused.

Python has a data type called a **byte string**; internally it's a sequence of byte values but it is displayed as a value that look like a string. But it is not a string. The table below summarizes them.

| Data Type | Description | Examples |
|----|----|----|
| String | Sequence of Unicode characters | "abc" or 'xyz' |
| Byte String | Sequence of bytes | b"abc" or b'xyz' |

Notice that byte strings are prefixed with the `b` character. Byte strings cannot contain artitrary Unicode characters; they can only contain characters that each fit into a single byte.

Byte strings are useful because they allow us, as programmers, to quickly define arrays of byte values but display them as things that look like strings. In this course there will be many opportunities to work with sequences of bytes. They are of a fixed size (8 bits only) and they encompass the ASCII character set which allows us to work with English text.

# Base 32 and Base 64 Encoding
The base 32 and base 64 encodings exist to convert artitrary sequences of bytes into a format that consists only of printable characters. When you attach an image or a PDF to an email, it must first be converted into a representation that an email message can accept. When the email standards were written in the 1980s, the authors did not forsee an age in which sending non-text would be a common usage. Therefore, emails can only contain printable ASCII characters. Any other data must be converted first.

Python provides some functions for encoding and decoding base 32 and base 64. They are part of the `base64` library, which you can import into your notebook with the following statement. **Execute the cell below to import the library.**

Then try the examples below of encoding and decoding base 32 and base 64.

In [ ]:
# Execute this cell to import the base64 encode/decode library
from base64 import *

In [41]:
b64encode(b'hello')

b'aGVsbG8='

In [42]:
b64encode(b'The River Runs Through It')

b'VGhlIFJpdmVyIFJ1bnMgVGhyb3VnaCBJdA=='

In [43]:
b64decode(b'TWF5IHRoZSBGb3JjZSBiZSB3aXRoIHlvdS4=')

b'May the Force be with you.'

In [44]:
b32encode(b'Sierra College')

b'KNUWK4TSMEQEG33MNRSWOZI='

In [45]:
b32decode(b'KJSW2ZLNMJSXELBAOJSW2ZLNMJSXEIDUNBSSANLUNAQG6ZRAJZXXMZLNMJSXE===')

b'Remember, remember the 5th of November'

### Your Turn!
Decode the following strings:

b'RG8gb3IgZG8gbm90OyB0aGVyZSBpcyBubyB0cnk='

b'JRUWMZJANFZSA53IMF2CA2DBOBYGK3TTEB3WQZLOEB4W65JHOJSSAYTVON4SA3LBNNUW4ZZAN52GQZLSEBYGYYLOOMXA===='

b'QWx3YXlzIHJlbWVtYmVyIHRoYXQgeW91IGFyZSBhYnNvbHV0ZWx5IHVuaXF1ZS4gSnVzdCBsaWtlIGV2ZXJ5b25lIGVsc2Uu'

In [19]:
from base64 import *
b64decode(b'RG8gb3IgZG8gbm90OyB0aGVyZSBpcyBubyB0cnk=')

b'Do or do not; there is no try'

In [11]:
from base64 import *
b64decode(b'QWx3YXlzIHJlbWVtYmVyIHRoYXQgeW91IGFyZSBhYnNvbHV0ZWx5IHVuaXF1ZS4gSnVzdCBsaWtlIGV2ZXJ5b25lIGVsc2Uu')

b'Always remember that you are absolutely unique. Just like everyone else.'

In [20]:
from base64 import *
b64decode(b'JRUWMZJANFZSA53IMF2CA2DBOBYGK3TTEB3WQZLOEB4W65JHOJSSAYTVON4SA3LBNNUW4ZZAN52GQZLSEBYGYYLOOMXA====')

b'%\x15\x161\x92@4VR\x03\x9d\xc80]\x82\x03`\xc18\x16\x06+t\xd3\x10\x1d\xd6A\x92\xce\x10\x1e\x16\xeb\x92G8\x94\x92\x01\x84\xd58\xde\x12\x03r\xc14\xd5\x16\xe1\x96@7\x9d\x86A\x92\xd2\x10\x16\x06a\x82\xce8\xc5\xc0'